### ISSUE TITLE
Eraser function suggestion

### ISSUE THREAD

**Issue by esaribeiro** [bigbluebutton/bigbluebutton#6026]
Hello, everyone!

I'm a BBB user and would like contribute with a sugestion.

I'd like to sugest to add an eraser button into whiteboard sidebar. 
I think the eraser function could be similar the pencil function, but with automatic color of the background.
For example:
When the user select de eraser option, the BBB apply the same of pencil function, but the color of brush should be the same (identified automatically) of the background area color.

**Comment by oswaldoacauan** [comment]
This can be easily done with SVG `<mask>`s. We just need to mimic the `pencil` as @esaribeiro suggested and use it as a mask for the slide.

```
<g>
  <mask id="eraser-mask-1">
    <rect x="0" y="0" width="100%" height="100%" fill="black"></rect>
    <path fill="white" stroke="#fff" d="..." stroke-width="69.63730569948187" 
    stroke-linejoin="round" stroke-linecap="round"></path>
  </mask>
  <use href="#image-slide" mask="url(#eraser-mask-1)"></use>
</g>```

Not sure how hard it would be in Flash since they use `swf` and not `svg` cc @capilkey

**Comment by capilkey** [comment]
I don't know of any mask type with Flash graphics. The recordings are also another consideration when trying to handle erasers.

**Comment by oswaldoacauan** [comment]
cc @kepstin

**Comment by oswaldoacauan** [comment]
I tried it on the developer tools itself and it seems to work as expected.

> ![image](https://user-images.githubusercontent.com/1268518/44927163-e5795a80-ad29-11e8-82d1-93537fb67b83.png)
> Flash client

> ![image](https://user-images.githubusercontent.com/1268518/44927148-dc888900-ad29-11e8-99d3-ec08ef717647.png)
> HTML5 client with the black path as a `<mask>`

**Comment by esaribeiro** [comment]
Hi,
was this functionality add to bbb?

Thanks

**Comment by ffdixon** [comment]
We haven't added this functionality to BigBlueButton, but it is on our road map.  We'll update this issue when we start working on it.  For more information on how we prioritize features, see http://docs.bigbluebutton.org/support/faq.html#when-will-feature-x-be-implemented

**Comment by gittester2017** [comment]
@oswaldoacauan , I tried the folowing code in the function render() of pencil component.jsx, it didn't work.
      <g>
        <mask id="eraser-mask-1">
          <rect x="0" y="0" width="100%" height="100%" fill="black"/>
          <path fill="white" stroke="#fff" d={this.getCurrentPath()} strokeWidth="69.63730569948187"
                strokeLinejoin="round" strokeLinecap="round"/>
        </mask>
        <use xlinkHref={currentSlide.svgUri} mask="url(#eraser-mask-1)"/>
      </g>
1. "d" is the path of pencil
2. xlinkHref is the url of first default presentation
Could you please show me more details about "mask" method? 
My own method is use path or clipPath to draw many circle on the path. It seems your "mask" method is better.

Thansks.
BestRegards.

**Comment by abegunasekara** [comment]
Hey Can someone please tell me how to add this tool to my BBB, (Noob)

**Comment by ffdixon** [comment]
We have plans to implement it in 2020.  It's going to take some work (which is why it's not been implemented already).

**Comment by abegunasekara** [comment]
Hey ffdixon, 
Thanks for the reply, Anyways can you please give a possible timeframe, Plus can u tell me how to use the above code. like where to edit and all. I use GCP with SSH

Thanks.

**Comment by gittester2017** [comment]
The above code is not enough, replay code also need update

**Comment by abegunasekara** [comment]
Did you manage to find a solution?

**Comment by ffdixon** [comment]
The eraser feature is something we (the core developers) are going to implement this year (in 2020).   For more information on how we prioritize features, see

  http://docs.bigbluebutton.org/support/faq.html#when-will-feature-x-be-implemented

**Comment by hippylover** [comment]
Interested in this too!

**Comment by stathisstefanidis1977** [comment]
I think it is essential that there is a rubber

**Comment by testprepturkey** [comment]
+1

**Comment by timokoesters** [comment]
Did you think about making it delete lines instead of behaving like a brush?

**Comment by BernhardBaumrock** [comment]
Mabe obvious, but maybe not (for me it was not, that's why I found this issue which brought me the idea):

On a white sheet you can just use the pencil tool whith white color to "erase" the parts you do not want. Or you can use the rectangle tool as well. Would be great to be able to choose the background color of the rectangle and circle etc. too, not only the stroke color :)

**Comment by matiasilva** [comment]
While that's a "quick fix", it's non-ideal. The strokes are all stored in memory and adding more strokes to replace the previous ones is wasted rendering time and inefficient.

**Comment by hiroshisuga** [comment]
Using svg's clippath, we will be able to implement a simple (and probably less heavy) eraser tool.

**Comment by hiroshisuga** [comment]
It seems that I have managed. This simply erases a rectangle area, but I believe it's sufficient. I will post a PR, in case somebody is interested in.

![screen](https://user-images.githubusercontent.com/45039819/101860137-78afc480-3bb0-11eb-84bc-58e56a9e7f1e.gif)

**Comment by stathisstefanidis1977** [comment]
How can we integrate this in our side ?

**Comment by hiroshisuga** [comment]
> How can we integrate this in our side ?

Everything is nicely described here:
https://docs.bigbluebutton.org/2.2/dev.html
Good luck!

**Comment by dkm062** [comment]
> It seems that I have managed. This simply erases a rectangle area, but I believe it's sufficient. I will post a PR, in case somebody is interested in.
> 
> ![screen](https://user-images.githubusercontent.com/45039819/101860137-78afc480-3bb0-11eb-84bc-58e56a9e7f1e.gif)
> 
> Another one to confirm an opaque drawing over an eraser.
> 
> ![screen](https://user-images.githubusercontent.com/45039819/101861299-0b516300-3bb3-11eb-9a83-d6c27b46c076.gif)


Will you please share the code you did to add an eraser in current BBB version.
I am having latest version of BBB and wanted to add this rubber function in the server

**Comment by hiroshisuga** [comment]
@dkm062 , absolutely! Please look at the link to a PR above.

**Comment by dcr18** [comment]
how can integrate this  Eraser function in our side please provide me code as soon as possible?

**Comment by hiroshisuga** [comment]
> how can integrate this Eraser function in our side please provide me code as soon as possible?

Everything is nicely described here:
https://docs.bigbluebutton.org/2.2/dev.html
Good luck!

**Comment by suprobhat** [comment]
hi hiroshisuga
my existing bbb-server I will not be changing the code. I will be requesting for a proper guideline I will add you eraser the function, please help me I will beg you.
Thanks for your reply 
Everything is nicely described here:
https://docs.bigbluebutton.org/2.2/dev.html
and here is the code:
#11018
please help me
any related dockment please provide me my personal gmail id : suprobhatganguly1997@gmail.com
and any related video please provide me .. :) :) 👍

**Comment by hiroshisuga** [comment]
Hello @suprobhat , I have to say it's not possible to add the eraser without touching the code. Perhaps you'd better wait for the PR being merged. Sorry that I cannot help you more..

**Comment by suprobhat** [comment]
ok thank you

**Comment by hiroshisuga** [comment]
Hi @suprobhat , I am afraid that you need to be the administrator of your server, and have the basic knowledge of Linux command line operation. If you have them, it wouldn't be very difficult to follow the instruction. Otherwise you will be more response and help at https://groups.google.com/g/bigbluebutton-dev . If you have more specific/technical questions, please come back and ask at #11018 .

**Comment by suprobhat** [comment]
Hi,
@hiroshisuga Download the BBB files in my Server, Make necessary changes in my SERVER. 
Now i get the modified files in my SERVER, Please help me to run the BBB install file in my server without taking BBB repository calling.

Here i'm attaching the error file. Kindly check and revert the solution.

![Error](https://user-images.githubusercontent.com/61926568/108694475-1a468100-7525-11eb-90af-19964678fb3c.jpg)

The script in run ./build_script.sh and chmod +x build_script.sh

**Comment by hiroshisuga** [comment]
Hi, you can follow the instruction here. Repository calling is not necesaary if you are not commiting your change to the base code.
https://docs.bigbluebutton.org/2.2/dev.html

**Comment by timmwille** [comment]
How is the status with this? just had a nice discussion about this and found the issue from 2018

**Comment by antobinary** [comment]
Erasing is included in BigBlueButton 2.6 (beta.2 stage as of this moment)

**Comment by dhhg12x** [comment]
Photoshop’s eraser serves as one of the program‘s most useful features.


**[Eraser Tool Not Working](https://btecz.com/eraser-tool-not-working-photoshop/)**
